feat: step 3 of `make transfer`

rootulp opened the pull request:
Implement step 3 of https://github.com/celestiaorg/celestia-zkevm-ibc-demo/issues/31

Closes https://github.com/celestiaorg/celestia-zkevm-ibc-demo/issues/138 b/c removes mock SP1 programs
Closes https://github.com/celestiaorg/celestia-zkevm-ibc-demo/issues/157 b/c removes mock SP1 programs
Closes https://github.com/celestiaorg/celestia-zkevm-ibc-demo/issues/180 b/c uses real SP1 programs from solidity-ibc-eureka
Closes https://github.com/celestiaorg/celestia-zkevm-ibc-demo/issues/204 b/c we only need 1

## Description

- Removes the need to use a fork of ibc-solidity-eureka
- Removes the need to copy + paste a diff of the `justfile`
- Measures the time it takes to generate proofs
- Refactors a lot of the client setup

## Tasks

- [x] Remove the README copy + paste of verifier key
- [x] Switch from fork of solidity-ibc-eureka to commit in upstream: https://github.com/cosmos/solidity-ibc-eureka/commit/8dad0c2502a8f5bc9cc970363eae115667f34754
- [x] Delete all the code for celestia-prover mock proofs b/c the real proofs are fast enough
- [x] Close all the issues related to celestia-prover mocks
- [x] Throw an error and stop `make transfer` if eth tx status 0 is hit.
- [x] Add a step before `make transfer` starts that checks the verifier key on the Tendermint light client == the verifier key on celestia-prover
- [x] Uncomment `_verifySP1Proof` in the Tendermint light client contract
- [x] Document the `.env` file that we need in solidity-ibc-eureka
- [x] Override the justfile to create genesis file with operator `--proof-type groth16`
- [x] Create a PR upstream to add verifier to deployed output JSON
- [x] Update the height used in prove membership to use the latest trusted height on the light client
- [x] Figure out why the SimApp is returning a value of 0 for the abci_query
- [x] Uncomment the kvValue check inside the SP1 program
- [x] Fix RecvPacket

## FLUPs

1. Remove verbose logging
2. Query balances before and after the transfer

rootulp commented:
I'm working on this.
- We're basically doing [this step](https://github.com/cosmos/solidity-ibc-eureka/blob/b241dc7658dc1540013430851f6e861e2df7cd51/e2e/interchaintestv8/ibc_eureka_test.go#L823) from the ibc_eureka_test.
- Under the hood that appears to call a relayer running via interchaintest. The relayer implementation is [here](https://github.com/cosmos/solidity-ibc-eureka/blob/6bb8fcf6af5094487c85f12d9398c8401fd4a1b7/programs/relayer/src/modules/eth_to_cosmos.rs#L133-L191) so we need to port that from Rust to Go

rootulp commented:
An event attribute is missing from the send packet: `packet_source_client`.

It is defined here: https://github.com/cosmos/ibc-go/blob/c0456f8d0aa31c1a218353a422d3311a57f3aaa4/modules/core/04-channel/v2/types/events.go#L17

which doesn't exist in the ibc-go version that SimApp uses: https://github.com/cosmos/ibc-go/blob/a9bdb8a50a3904bdf0589690321e153f121d824a/modules/core/04-channel/types/events.go

rootulp commented:
We can temporarily hack around it by hard-coding the expected `packet_source_client` and then in a FLUP upgrade ibc-go to the version that solidity-ibc-eureka uses.

ninabarbakadze commented on testing/demo/pkg/relay/main.go:
```diff
@@ -209,3 +225,326 @@ func getUpdateClientArguments() (abi.Arguments, error) {
+	// Load the ABI for the ICS26Router contract
+	// In a real implementation, we would have the ABI available
+	// For demonstration, create a simple ABI representation
+	recvPacketABI := `[{"inputs":[
```
out of curiosity: why don't you call RecvPacket on the contract directly? 
```
tx, err := ics26Contract.RecvPacket(getTransactOpts(faucet, eth), msgReceivePacket)
```

rootulp replied:
This was Cursor generated and I haven't gotten to this step. It should prob invoke the contract directly like you said.

rootulp commented:
Able to generate the membership proof now.

```
Packet sequence: 2
Generating proof for packet commitment with path: 30382d67726f746831362d30010000000000000002
Requesting celestia-prover state membership proof...
Received celestia-prover state membership proof with height 12039.
Created transaction: 0xaa48606de549ac43b04030363fdf626e007b7b0bfce6a680c21249275a759384
2025/03/05 11:23:39 Failed to relay transaction: failed to send transaction: already known
```

rootulp commented:
EVM error:

```
"9fff831f0000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000000000000096368616e6e656c2d300000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000c30382d67726f746831362d300000000000000000000000000000000000000000"
```

which maps to

```
| Error | IBCInvalidCounterparty(string,string)        | 0x9fff831f |
```

rootulp commented:
New EVM error:

```
"df95155a000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000096368616e6e656c2d300000000000000000000000000000000000000000000000"
```

which maps to

```
| Error | IBCCounterpartyClientNotFound(string) | 0xdf95155a |
```

where the string is `channel-0`. I need to draw out the names of the things created in `make setup`

rootulp commented:
The reason we're hitting an error is because `make setup` creates the light client on the EVM roll-up with a counterparty clientId of `channel-0` even though it should be `08-groth16-0`.

rootulp commented:
Currently fails with

```
Queried transaction and extracted 16 events.
Extracted SendPacket event from transaction: map[encoded_packet_hex:0804120c30382d67726f746831362d301a0f30372d74656e6465726d696e742d3020abd8c7be062abc040a087472616e7366657212087472616e736665721a0769637332302d31221a6170706c69636174696f6e2f782d736f6c69646974792d6162692a8004000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000a000000000000000000000000000000000000000000000000000000000000000e00000000000000000000000000000000000000000000000000000000000000140000000000000000000000000000000000000000000000000000000000000006400000000000000000000000000000000000000000000000000000000000001a000000000000000000000000000000000000000000000000000000000000000057374616b65000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002d636f736d6f73316c74767a7077663365673865397337777a6c6571646d7730326c657372646578396a6774307100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002a30783766333963353831663539356235336335636231396235613665356238663361306231663266366500000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000d74657374207472616e7366657200000000000000000000000000000000000000 msg_index:0 packet_dest_client:07-tendermint-0 packet_sequence:4 packet_source_client:08-groth16-0 packet_timeout_timestamp:1741810731]
Packet sequence: 4
Generating proof for packet commitment with path: 30382d67726f746831362d30010000000000000004
Requesting celestia-prover state membership proof...
Received celestia-prover state membership proof with height 360.
sendPacketEvent: map[encoded_packet_hex:0804120c30382d67726f746831362d301a0f30372d74656e6465726d696e742d3020abd8c7be062abc040a087472616e7366657212087472616e736665721a0769637332302d31221a6170706c69636174696f6e2f782d736f6c69646974792d6162692a8004000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000a000000000000000000000000000000000000000000000000000000000000000e00000000000000000000000000000000000000000000000000000000000000140000000000000000000000000000000000000000000000000000000000000006400000000000000000000000000000000000000000000000000000000000001a000000000000000000000000000000000000000000000000000000000000000057374616b65000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002d636f736d6f73316c74767a7077663365673865397337777a6c6571646d7730326c657372646578396a6774307100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002a30783766333963353831663539356235336335636231396235613665356238663361306231663266366500000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000d74657374207472616e7366657200000000000000000000000000000000000000 msg_index:0 packet_dest_client:07-tendermint-0 packet_sequence:4 packet_source_client:08-groth16-0 packet_timeout_timestamp:1741810731]
sendPacketEvent[packet_timeout_timestamp]: 1741810731
Using payload version: ics20-1, encoding: application/x-solidity-abi
Created transaction: 0xeea338c20413601aeb51392b7f6bc8755a472f74b1514a4d94198ad8449d5812
2025/03/12 15:58:20 Failed to relay transaction: failed to send transaction: already known
exit status 1
make: *** [transfer] Error 1
```

rootulp commented:
I think it means the Ethereum tx has already been seen by Reth but idk how or why b/c I'm crafting a new tx. It may be because the nonce that's used is old? Investigating...

rootulp commented:
```
Created transaction with hash: 0x733252007e7c95bc644c9d5ee52652d0c8302ebffea6e5ee00cd181d0beb5e20 and nonce: 21
2025/03/18 16:15:01 Failed to relay IBC transaction: transaction already pending in mempool
```

rootulp commented:
The RecvPacket tx on EVM is failing

```
Created transaction with hash: 0x414f80075dabe83fb0673584a4834ec064225946fe7e0c3de9312fa738277fcd and nonce: 33
Transaction failed. Inspect logs or contract.
Transaction sent to Ethereum, hash: 0x414f80075dabe83fb0673584a4834ec064225946fe7e0c3de9312fa738277fcd, status: 0
```

I got the EVM trace and it seems like it's failing on this line

```
{"pc":71,"op":"DELEGATECALL","gas":2962156,"gasCost":2915913,"depth":1,"error":"CallOrCreate","stack":["0x0","0x0","0x744","0x0","0xdcea98814d56e04dd8b316c79d62efd49bc34840","0x2d32ec"]
```

There is a contract deployed to that address because I can see it's bytecode.

```
cast code 0xdcea98814d56e04dd8b316c79d62efd49bc34840 > cast-code.txt
```

I'm looking at /Users/rootulp/git/rootulp/celestiaorg/celestia-zkevm-ibc-demo/solidity-ibc-eureka/broadcast/E2ETestDeploy.s.sol/80087/run-latest.json to see what that contract is supposed to do because it's not in the map of deployed contract addresses.

rootulp commented:
![proxy](https://github.com/user-attachments/assets/eca3882e-9343-4c69-8fbb-89dd3acccfad)

rootulp commented:
The function selectors on the router contract:

```
0x01ffc9a7	bytes4                                                       	pure
0x075beb64	address                                                      	nonpayable
0x0bb1774c	bytes,uint64                                                 	pure
0x1de3d58f	address                                                      	nonpayable
0x1ec43e23	bytes,uint256,address                                        	view
0x2447af29	bytes                                                        	view
0x248a9ca3	uint256                                                      	view
0x24d7806c	address                                                      	view
0x258ef1b3	(uint256,uint256,bytes,uint256,uint256[])                    	view
0x27f146f3	                                                             	view
0x2f2ff15d	uint256,address                                              	nonpayable
0x340cbac4	address                                                      	nonpayable
0x365388a2	                                                             	view
0x36568abe	uint256,address                                              	nonpayable
0x4215e843	bytes,address                                                	view
0x4415585c	bytes,address                                                	view
0x485cc955	address,address                                              	nonpayable
0x4b720d5b	address                                                      	pure
0x4d6e7ce3	(uint256,uint256,bytes)                                      	nonpayable
0x4f1ef286	address,bytes                                                	payable
0x52d1902d	                                                             	pure
0x54a5979b	                                                             	view
0x5f516889	bytes,address                                                	view
0x6e65a7d0	(uint256,bytes,uint256,uint256,uint256[])                    	view
0x6fbf8079	bytes,bytes                                                  	nonpayable
0x7795820c	uint256                                                      	view
0x7eb78932	bytes                                                        	view
0x7ecfb0c2	(uint64,bytes,bytes,uint64,(bytes,bytes,bytes,bytes,bytes)[])	pure
0x805cc564	address                                                      	nonpayable
0x9046a3a1	address                                                      	nonpayable
0x91d14854	uint256,address                                              	view
0x9e2e5c83	bytes,bytes                                                  	nonpayable
0x9ea7ff99	bytes,bytes                                                  	pure
0xa217fddf	                                                             	pure
0xac9650d8	bytes[]                                                      	pure
0xad3cb1cc	                                                             	pure
0xb0777bfa	bytes                                                        	view
0xb0830ab9	bytes                                                        	pure
0xd3352436	                                                             	pure
0xd547741f	uint256,address                                              	nonpayable
0xdb8b4b63	bytes,uint64                                                 	pure
0xdf5426a2	                                                             	pure
0xdfdcf16d	bytes,uint64                                                 	pure
0xe0490029	(uint256,bytes,uint256,uint256,uint256[])                    	view
0xe3cb36a0	uint256,address                                              	nonpayable
0xede05f16	(uint64,bytes,bytes,uint64,(bytes,bytes,bytes,bytes,bytes)[])	pure
0xf4b1b9c6	bytes,bytes                                                  	nonpayable
0xf6d1d9c5	address                                                      	nonpayable
```

I assume the failing function selector maps to RecvPacket so looking into if we may be calling that incorrectly

rootulp commented:
Got the solidity-ibc-eureka e2e test to run: [e2e.log](https://github.com/user-attachments/files/19349874/e2e.log). Attaching files used for debugging:
[cast-code.txt](https://github.com/user-attachments/files/19349890/cast-code.txt)
[debug-txhash-1.json](https://github.com/user-attachments/files/19349891/debug-txhash-1.json)
[debug-txhash-2.json](https://github.com/user-attachments/files/19349892/debug-txhash-2.json)
[run-latest.json](https://github.com/user-attachments/files/19349893/run-latest.json)

rootulp commented:
I was able to get eureka e2e test passing: https://github.com/cosmos/solidity-ibc-eureka/issues/428 but the steps there broke the update client step in this repo 😕

rootulp commented:
The update client ABI hasn't changed so I'm reasonably confident it's being invoked correctly.

```json
{
    "type": "function",
    "name": "updateClient",
    "inputs": [
      {
        "name": "clientId",
        "type": "string",
        "internalType": "string"
      },
      {
        "name": "updateMsg",
        "type": "bytes",
        "internalType": "bytes"
      }
    ],
    "outputs": [
      {
        "name": "",
        "type": "uint8",
        "internalType": "enum ILightClientMsgs.UpdateResult"
      }
    ],
    "stateMutability": "nonpayable"
  },
```

I may have found a way to emit logs from inside the contract so I'm attempting to that

rootulp commented:
Confirmed the router contract is correct, the tendermint contract address is correct, the verifier key is correct.

rootulp commented:
I'm testing a hypothesis: the mock programs may be broken so I'm trying to use the real update client SP1 program from solidity-ibc-eureka instead of our mock.

rootulp commented:
When using the correct verifier key, I hit the same panic again
```
4e487b710000000000000000000000000000000000000000000000000000000000000001
```

rootulp commented:
I fixed the update client issue by downgrading my local `operator` installation to from a previous solidity-ibc-eureka commit:
so I expect the `trustedClientState` or `trustedConsensusState` generated by the latest operator binary to be invalid. For now: use the fixed-genesis.json from

Ref: https://gist.github.com/rootulp/dd0e4ef48b915e06e658bb1323e7c910

rootulp commented:
Now debugging errors in the membership proof creation b/c now we use the real proofs from solidity-ibc-eureka

rootulp commented:
```
Got state membership request...
Generating proof with key_proofs: [([[105, 98, 99], [51, 48, 51, 56, 50, 100, 54, 55, 55, 50, 54, 102, 55, 52, 54, 56, 51, 49, 51, 54, 50, 100, 51, 48, 48, 49, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 51]], [], MerkleProof { proofs: [CommitmentProof { proof: Some(Nonexist(NonExistenceProof { key: [51, 48, 51, 56, 50, 100, 54, 55, 55, 50, 54, 102, 55, 52, 54, 56, 51, 49, 51, 54, 50, 100, 51, 48, 48, 49, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 51], left: Some(ExistenceProof { key: [48, 56, 45, 103, 114, 111, 116, 104, 49, 54, 45, 48, 1, 0, 0, 0, 0, 0, 0, 0, 2], value: [135, 133, 55, 245, 129, 73, 97, 59, 205, 126, 15, 19, 36, 151, 90, 151, 119, 81, 159, 192, 132, 96, 17, 217, 201, 127, 191, 182, 21, 22, 242, 145], leaf: Some(LeafOp { hash: Sha256, prehash_key: NoHash, prehash_value: Sha256, length: VarProto, prefix: [0, 2, 194, 1] }), path: [InnerOp { hash: Sha256, prefix: [2, 4, 194, 1, 32, 7, 79, 148, 177, 205, 25, 178, 176, 132, 115, 3, 92, 75, 117, 59, 80, 91, 244, 203, 117, 187, 151, 1, 39, 26, 76, 23, 207, 13, 161, 2, 220, 32], suffix: [] }, InnerOp { hash: Sha256, prefix: [4, 8, 194, 1, 32], suffix: [32, 219, 128, 254, 92, 145, 212, 28, 247, 219, 108, 240, 92, 239, 238, 126, 72, 19, 101, 139, 60, 69, 177, 106, 231, 228, 13, 244, 179, 170, 140, 151, 158] }, InnerOp { hash: Sha256, prefix: [6, 16, 194, 1, 32], suffix: [32, 103, 235, 234, 79, 7, 23, 213, 13, 233, 205, 147, 154, 21, 93, 236, 62, 215, 67, 214, 161, 62, 202, 233, 24, 218, 36, 183, 38, 250, 227, 138, 148] }, InnerOp { hash: Sha256, prefix: [8, 30, 194, 1, 32], suffix: [32, 36, 139, 188, 106, 200, 11, 126, 211, 205, 141, 90, 128, 70, 233, 115, 237, 165, 201, 17, 133, 159, 159, 46, 63, 10, 89, 87, 116, 225, 205, 54, 43] }] }), right: Some(ExistenceProof { key: [99, 108, 105, 101, 110, 116, 80, 97, 114, 97, 109, 115], value: [10, 1, 42], leaf: Some(LeafOp { hash: Sha256, prehash_key: NoHash, prehash_value: Sha256, length: VarProto, prefix: [0, 2, 2] }), path: [InnerOp { hash: Sha256, prefix: [2, 4, 194, 1, 32], suffix: [32, 20, 196, 167, 85, 139, 2, 166, 209, 244, 130, 244, 71, 170, 197, 143, 134, 188, 63, 54, 223, 113, 146, 131, 239, 40, 5, 228, 242, 207, 50, 217, 138] }, InnerOp { hash: Sha256, prefix: [4, 8, 194, 1, 32, 110, 25, 199, 121, 74, 9, 72, 134, 227, 122, 205, 228, 35, 25, 8, 126, 185, 57, 215, 101, 121, 55, 255, 53, 243, 13, 224, 173, 161, 167, 12, 150, 32], suffix: [] }, InnerOp { hash: Sha256, prefix: [6, 16, 194, 1, 32], suffix: [32, 103, 235, 234, 79, 7, 23, 213, 13, 233, 205, 147, 154, 21, 93, 236, 62, 215, 67, 214, 161, 62, 202, 233, 24, 218, 36, 183, 38, 250, 227, 138, 148] }, InnerOp { hash: Sha256, prefix: [8, 30, 194, 1, 32], suffix: [32, 36, 139, 188, 106, 200, 11, 126, 211, 205, 141, 90, 128, 70, 233, 115, 237, 165, 201, 17, 133, 159, 159, 46, 63, 10, 89, 87, 116, 225, 205, 54, 43] }] }) })) }, CommitmentProof { proof: Some(Exist(ExistenceProof { key: [105, 98, 99], value: [111, 244, 81, 217, 182, 40, 74, 76, 121, 254, 244, 86, 12, 16, 190, 153, 112, 42, 90, 202, 128, 56, 232, 218, 188, 62, 202, 203, 153, 149, 14, 130], leaf: Some(LeafOp { hash: Sha256, prehash_key: NoHash, prehash_value: Sha256, length: VarProto, prefix: [0] }), path: [InnerOp { hash: Sha256, prefix: [1, 6, 185, 156, 13, 129, 25, 255, 30, 219, 203, 225, 101, 208, 241, 147, 55, 219, 188, 8, 14, 103, 124, 136, 229, 122, 162, 174, 118, 126, 191, 15, 15], suffix: [] }, InnerOp { hash: Sha256, prefix: [1, 144, 21, 110, 160, 64, 25, 127, 231, 35, 10, 146, 37, 233, 1, 233, 160, 203, 160, 154, 156, 40, 208, 133, 174, 130, 129, 194, 9, 212, 212, 65, 242], suffix: [] }, InnerOp { hash: Sha256, prefix: [1], suffix: [199, 66, 153, 185, 71, 49, 63, 8, 167, 137, 92, 221, 88, 114, 235, 100, 72, 90, 83, 39, 143, 101, 224, 200, 223, 152, 93, 214, 220, 132, 130, 253] }, InnerOp { hash: Sha256, prefix: [1, 130, 167, 48, 124, 44, 55, 41, 212, 163, 12, 192, 164, 155, 247, 158, 238, 78, 238, 162, 233, 25, 120, 211, 19, 72, 126, 56, 22, 253, 255, 163, 80], suffix: [] }, InnerOp { hash: Sha256, prefix: [1], suffix: [50, 182, 198, 120, 169, 222, 122, 220, 220, 47, 99, 255, 77, 203, 189, 122, 243, 244, 39, 45, 60, 229, 222, 137, 58, 192, 203, 175, 132, 18, 179, 71] }] })) }] })]
stderr:
stderr: thread '<unnamed>' panicked at programs/sp1-programs/membership/src/main.rs:36:65:
stderr: called `Result::unwrap()` on an `Err` value: TypeCheckFail { expected_type: "offset (usize)", data: "020000000000000003000000000000006962632a000000000000003330333832" }
stderr: stack backtrace:
stderr: note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

rootulp commented:
Failing to update the Tendermint light client still. Most recent EVM tx hash `0x84fe0ad419b89caef5ce4c3292104fc00297a8f581c58006e7fbc8e41154f76f` with a revert reason of `4e487b710000000000000000000000000000000000000000000000000000000000000001`. I think it is because the client's `genesis.json` is messed up so going to try copy + pasting in a working one.

rootulp commented:
I'm re-deploying and testing with a commented out verify SP1 proof invocation in the tendermint contract.

rootulp commented:
UpdateClient works if I skip the verify SP1 proof step inside the Tendermint contract.

rootulp commented:
Trying after a rm -rf node_modules and bun install

rootulp commented:
Update: that didn't resolve. Still hit

```
4e487b710000000000000000000000000000000000000000000000000000000000000001
```

rootulp commented:
I'm going to skip the verify SP1 proof in the Tendermint light client to see if the remaining stuff works.

rootulp commented:
RecvPacket is also failing

[debug-recv-packet.json](https://github.com/user-attachments/files/19405568/debug-recv-packet.json)

rootulp commented:
When I tested the e2e test in solidity-ibc-eureka, I modified my `.env`. The updated version has

```env
VERIFIER="mock"
```

which deploys a mock verifier [here](https://github.com/cosmos/solidity-ibc-eureka/blob/8dad0c2502a8f5bc9cc970363eae115667f34754/scripts/E2ETestDeploy.s.sol#L61-L73). My hypothesis is that was causing the EVM panic so I'm re-deploying with 

```env
VERIFIER=""
```

so that it deploys a Groth16 verifier.

rootulp commented:
Found another issue, the operator by default creates a genesis.json with proof type plonk. We have to override it with a flag.

rootulp commented:
The iteration speed on debugging the RecvPacket is crazy slow b/c I tear down the stack and re-deploy it. To speed it up, consider using `MOCK` mode for the proof generation

rootulp commented:
The ibc_eureka_test calls RelayerClient.RelayByTx which is implemented in relayer/src/modules/cosmos_to_eth.rs.
It gets Cosmos tx events and plumbs them to tx_builder.relay_events which calls src_events_to_recv_and_ack_msgs. A MsgRecvPacket is composed of: packet, proof height, proof commitment (empty bytes). Then it later fills the proof commitment with an SP1 proof. When it injects the SP1 proof, it generates a commitment path from the send packet using:
1. source client
2. byte(1)
3. sequence in big endian
which is what we do. I think next step should be to try querying the SimApp to see if the packet commitment path we're generating points to a piece in state.

rootulp commented:
```shell
curl "http://localhost:5123/" \
--header "Content-Type: application/json" \
--data '{
  "jsonrpc": "2.0",
  "id": 1,
  "method": "abci_query",
  "params": {
  	"path": "/store/ibc/key",
    "data": "30382d67726f746831362d30010000000000000001",
    "height": "15",
    "prove": true
  }
}'
```

```json
{
  "jsonrpc": "2.0",
  "id": 1,
  "result": {
    "response": {
      "code": 0,
      "log": "",
      "info": "",
      "index": "0",
      "key": "MDgtZ3JvdGgxNi0wAQAAAAAAAAAB",
      "value": "2Ao6VPnQHrrbMSRNU/JSqVsYiygB71SA23mdvs3eJk8=",
      "proofOps": {
        "ops": [
          {
            "type": "ics23:iavl",
            "key": "MDgtZ3JvdGgxNi0wAQAAAAAAAAAB",
            "data": "CvoBChUwOC1ncm90aDE2LTABAAAAAAAAAAESINgKOlT50B662zEkTVPyUqlbGIsoAe9UgNt5nb7N3iZPGgsIARgBIAEqAwACHiIrCAESBAIEHiAaISBUXm+JxPF9FVbcYy75Q41ZIyisltH7dXHhg55Gu7djkiIrCAESBAQGHiAaISCY73o5or5wGvULRkV5rqOfCt1gJFYamzMMRO4cfzieCSIrCAESBAYOHiAaISDDke+uWoiw9ssA5iAowAccKhsRDX03xqWh/csZQGpneiIrCAESBAgcHiAaISDVC5wu5inx59OEIZSppzMq2RMCiBSFVXn3J6+zwHytLA=="
          },
          {
            "type": "ics23:simple",
            "key": "aWJj",
            "data": "CvkBCgNpYmMSIAuSnI1Y4dKj6MNszZmX7IyiXkS2Jkt7jgtSmhWwlEayGgkIARgBIAEqAQAiJQgBEiEBBrmcDYEZ/x7by+Fl0PGTN9u8CA5nfIjleqKudn6/Dw8iJQgBEiEBkBVuoEAZf+cjCpIl6QHpoMugmpwo0IWugoHCCdTUQfIiJwgBEgEBGiD5vBq135eJWUkc5pw+LnGs78mkAQ6Dkdo3bfO+Okjb6yIlCAESIQGzHhiVZ+Mk+5qrIO1dLy+yIUzN77cWXqwYvniLPXa97SInCAESAQEaINqAXbC/HrfHELURfSDX42oAXPMx7/63dm/UcJtbGmVp"
          }
        ]
      },
      "height": "15",
      "codespace": ""
    }
  }
}
```

rootulp commented:
https://github.com/cosmos/solidity-ibc-eureka/blob/8dad0c2502a8f5bc9cc970363eae115667f34754/contracts/light-clients/SP1ICS07Tendermint.sol#L253

rootulp commented:
Just hit a new error

```
5f1ca38100000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000000000080000000000000000000000000000000000000000000000000000000000000002005b45a664badf151ef5783b0ce665b3da63939278740f28047753819714d256e0000000000000000000000000000000000000000000000000000000000000000
```

```
| Error    | MembershipProofValueMismatch(bytes,bytes)               | 0x5f1ca381 |
```

rootulp commented:
Basically something is broken with respect to the key / value and proof generation.

rootulp commented:
I hit `fef760c7` MembershipProofKeyNotFound when I don't convert the path into hex.
I hit `5f1ca381` MembershipProofValueMismatch  when I do convert the path into hex.

rootulp commented:
Retrying after I comment out [this code](https://github.com/cosmos/solidity-ibc-eureka/blob/8dad0c2502a8f5bc9cc970363eae115667f34754/contracts/light-clients/SP1ICS07Tendermint.sol#L236-L254)

rootulp commented:
```
membership_output path: [0x696263, 0x333033383264363737323666373436383331333632643330303130303030303030303030303030303031] value: 0x
```

If you decode that from hex into ASCII you get: ["ibc", "30382d67726f746831362d30010000000000000001"]. If you decode the second element you get: 08-groth16-0 followed by 0x01 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x01 which is what we expect. It doesn't make sense that the value is empty.

rootulp commented:
The celestia-prover is finding showing that no value is returned after the abci_query for a path:

```
path: [[105, 98, 99], [51, 48, 51, 56, 50, 100, 54, 55, 55, 50, 54, 102, 55, 52, 54, 56, 51, 49, 51, 54, 50, 100, 51, 48, 48, 49, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 49]]
value: []
```

rootulp commented:
The value is zero for height 15 and non-zero for height 16.

```
{
  "jsonrpc": "2.0",
  "id": 1,
  "result": {
    "response": {
      "code": 0,
      "log": "",
      "info": "",
      "index": "0",
      "key": "MDgtZ3JvdGgxNi0wAQAAAAAAAAAB",
      "value": null,
      "proofOps": {
        "ops": [
          {
            "type": "ics23:iavl",
            "key": "MDgtZ3JvdGgxNi0wAQAAAAAAAAAB",
            "data": "Eu4BChUwOC1ncm90aDE2LTABAAAAAAAAAAEa1AEKDGNsaWVudFBhcmFtcxIDCgEqGgsIARgBIAEqAwACAiIrCAESBAIEECAaISAMTRVRgNTZrHT0N5T1L+LAuBp4SpsKqVtlJcMsR1utUSIrCAESBAQIECAaISAgH5kTs2NjrCjEIt/7EAQhWJT2rh6k5ttnAr4tnDHYVyIrCAESBAYMGCAaISBGYdqlpxTpqnLaIoIk5ijHeu1cEK3CP6rJIWxYZyvl+SIrCAESBAgaGCAaISCo5UQnlzPY0LwRZsEwvTOMLFATUGL/AmEfJRvZdETORQ=="
          },
          {
            "type": "ics23:simple",
            "key": "aWJj",
            "data": "CvkBCgNpYmMSIMfJs0z9DHdQQvFACIqiKZLD1Dqbi+G8tFqSl1kDpv/7GgkIARgBIAEqAQAiJQgBEiEBBrmcDYEZ/x7by+Fl0PGTN9u8CA5nfIjleqKudn6/Dw8iJQgBEiEBkBVuoEAZf+cjCpIl6QHpoMugmpwo0IWugoHCCdTUQfIiJwgBEgEBGiDS6UX5/3nqr8uu1kjdaWQzvHwTAZTKUyoOEzeOoQG8rSIlCAESIQHaY5A0iSBLtJ63UvBYRjUh7rReu4TDvqpdz5U3qpqeTyInCAESAQEaICMbLMZsal7Y6iVQhqdU9EdoeB1VZB7Iv/a5+urdiaa7"
          }
        ]
      },
      "height": "15",
      "codespace": ""
    }
  }
}
```


```
curl "http://localhost:5123/" \
--header "Content-Type: application/json" \
--data '{
  "jsonrpc": "2.0",
  "id": 1,
  "method": "abci_query",
  "params": {
  	"path": "/store/ibc/key",
    "data": "30382d67726f746831362d30010000000000000001",
    "height": "16",
    "prove": true
  }
}'

{
  "jsonrpc": "2.0",
  "id": 1,
  "result": {
    "response": {
      "code": 0,
      "log": "",
      "info": "",
      "index": "0",
      "key": "MDgtZ3JvdGgxNi0wAQAAAAAAAAAB",
      "value": "mIOJD571i4z2T0nm+5qG1sNw+TPJY0fb3VGo01oRml0=",
      "proofOps": {
        "ops": [
          {
            "type": "ics23:iavl",
            "key": "MDgtZ3JvdGgxNi0wAQAAAAAAAAAB",
            "data": "CvoBChUwOC1ncm90aDE2LTABAAAAAAAAAAESIJiDiQ+e9YuM9k9J5vuahtbDcPkzyWNH291RqNNaEZpdGgsIARgBIAEqAwACICIrCAESBAIEICAaISBUXm+JxPF9FVbcYy75Q41ZIyisltH7dXHhg55Gu7djkiIrCAESBAQGICAaISAMTRVRgNTZrHT0N5T1L+LAuBp4SpsKqVtlJcMsR1utUSIrCAESBAYOICAaISDJrgAuYgExLaOmpbAchQoc6O9HKg+5biOyrATxQ6fxXiIrCAESBAgcICAaISCX40mtD9sLnYL3YVATJCgc6pLgiQSdDpag1k/AZKPG6A=="
          },
          {
            "type": "ics23:simple",
            "key": "aWJj",
            "data": "CvkBCgNpYmMSINivEGheJSQya1ohsMwTHEF+vbJFza3prQykQ5yi8S/HGgkIARgBIAEqAQAiJQgBEiEBBrmcDYEZ/x7by+Fl0PGTN9u8CA5nfIjleqKudn6/Dw8iJQgBEiEBkBVuoEAZf+cjCpIl6QHpoMugmpwo0IWugoHCCdTUQfIiJwgBEgEBGiCloM2GhUNfYhSu9Qa9kNrYvtHc58p2H1/cE1ibG1FvqSIlCAESIQH4/gLMfn1jPab+huv6MH3T7wrij1ZSuLAS+gxPnryR8CInCAESAQEaIE9VSE/Z7aKsBJO7SpgnQ0CFmDtsWblKmR+zjaHUXDjM"
          }
        ]
      },
      "height": "16",
      "codespace": ""
    }
  }
}
```

The prove_path implementation is querying one before the block height that I care about because the `- 1` in

```rust
    async fn prove_path(&self, path: &[Vec<u8>], height: u32) -> Result<(Vec<u8>, MerkleProof)> {
        let res = self
            .abci_query(
                Some(format!("store/{}/key", std::str::from_utf8(&path[0])?)),
                path[1..].concat(),
                // Proof height should be the block before the target block.
                Some((height - 1).into()),
                true,
            )
            .await?;
```

rootulp commented:
Now celestia-prover generates a proof with a value. But we're hitting

```
5f1ca381
```

The mismatch is:

```
0x4162 != 0xb8d6c79e26df3dd0eac3e89250e36f1e4b4d818bbf3183469ebfbe7b80ef3bb5
```

The second value does match what the celestia-prover value fetched. IDK where the first value is coming from yet

rootulp commented:
I think the Solidity contracts are creating the packet commitment bytes incorrectly because isn't even 32 bytes and it should be according to https://github.com/cosmos/solidity-ibc-eureka/blob/8dad0c2502a8f5bc9cc970363eae115667f34754/contracts/ICS26Router.sol#L177

rootulp commented:
If I comment out the kvPath comparison, then I hit

```
ConsensusStateNotFound
```

rootulp commented:
I wanted to add an assertion at the top of relayByTx that checks the latest trusted height on the light client > source tx height but I can't do it because it's a pain to get the trusted height inside Go code. There are no Go utilities to convert the client state bytes into a client state struct and it is ABI encoded. So I refactored so that the celestia-prover fetches the latest trusted client height in Rust code.

rootulp commented:
New error unlocked

```
| Error    | IBCFailedCallback()                                                                                              | 0xadef7fb8                                                         
```

rootulp commented:
I was just able to relay a packet 🎊

rootulp commented:
It actually finishes relaying the packet!

```
RecvPacket success in block 276
Relayed IBC transaction 584C86FB1FA51AC36E6F36CF41B217CDCCA152FCF87446BD08713A50DB8C3386 to client 07-tendermint-0...
```

cmwaters commented on provers/celestia-prover/README.md:
````diff
@@ -64,14 +64,11 @@ Push new image to GHCR:
-# Regenerate mock-update-client-elf
-$ ~/.sp1/bin/cargo-prove prove build --elf-name mock-update-client-elf --output-directory provers/celestia-prover/elf/ --packages mock-update-client --no-default-features
-```
+| Proof            | Time       | Cycles |
+|------------------|------------|--------|
+| State Transition | 80s - 235s | TBD    |
+| Membership       | 105s       | TBD    |
````
👍 

Is the variance dependent on the validator set - are there any other factors that would contribute to generation time

rootulp replied:
Discussed on Slack. I think the variance is based on which machine in the prover networks services the request b/c the demo has the same SimApp / validator set / and number of blocks each time I create a proof and there is variance.

cmwaters commented on provers/celestia-prover/prover/src/main.rs:
```diff
@@ -176,18 +230,65 @@ impl Prover for ProverService {
+            );
+        });
+
+        let membership_proof = MembershipProof::from(SP1MembershipProof {
+            sp1Proof: SP1Proof::new(
+                &self.membership_prover.vkey.bytes32(),
+                sp1_proof.bytes(),
+                sp1_proof.public_values.to_vec(),
+            ),
+            trustedConsensusState: trusted_consensus_state,
+        });
```
Wow did you get the zk membership proofs to work! nice job!

rootulp replied:
Yes. TBH it was way easier to use the SP1 programs written in solidity-ibc-eureka than to debug the issues in our mock SP1 programs.

cmwaters commented on testing/demo/pkg/setup/groth16.go:
```diff
@@ -4,70 +4,74 @@ import (
+		return nil, nil, err
+	}
+
+	// TODO: Query the stateTransitionVerifierKey and stateMembershipVerifierKey from the EVM prover.
+	// Query the celestia prover info endpoint for the state transition verifier key
+	// conn, err := grpc.NewClient(evmProverRPC, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	// if err != nil {
+	// 	return nil, nil, fmt.Errorf("failed to connect to prover: %w", err)
+	// }
+	// defer conn.Close()
```
How come this was commented out, don't we need the verifier keys

rootulp replied:
Yes the Groth16 client needs the verifier keys but it was never set up with them. I started implementing in this PR but it's not necessary for step 3. This only became possible to implement recently b/c now there's an EVM-prover Docker service.

A TODO after this merges is to uncomment this code.

cmwaters commented on testing/demo/pkg/setup/tendermint.go:
```diff
@@ -0,0 +1,114 @@
+
+// deployEurekaContracts deploys all of the IBC Eureka contracts (including the
+// SP1 ICS07 Tendermint light client contract) on the EVM roll-up.
+func deployEurekaContracts() error {
```
Q: How are the values for initialising specifically the sp1 tendermint light client contract set?

rootulp replied:
They're read from `solidity-ibc-eureka/scripts/genesis.json` which we had a bunch of issues with in the past b/c the deployment commands previously would invoke a `just` command to generate that file and then modify it with verifier keys for the hollowed out SP1 programs.

Now the `makefile` in our repo creates the genesis.json directly using the `operator` binary. It sets the correct verifier keys.

cmwaters commented on testing/demo/pkg/setup/tendermint.go:
```diff
@@ -0,0 +1,114 @@
+// the EVM roll-up.
+//
+// Note this also registers the counterparty.
+func addClientToRouter(addresses utils.ContractAddresses) error {
```
Didn't realise there needed to be a transaction to link the two smart contracts

rootulp replied:
Yep. Adds it here: https://github.com/rootulp/celestia-zkevm-ibc-demo/blob/80d92ce1393a523173aac577220f5b38563aa26f/testing/demo/pkg/setup/tendermint.go#L97

cmwaters commented on testing/demo/pkg/transfer/celestia_prover_utils.go:
```diff
@@ -0,0 +1,57 @@
+	return key, nil
+}
+
+func getCelestiaProverInfo() (*proverclient.InfoResponse, error) {
```
I guess as a flup we should provider the address as an argument because this function can be used for both the celestia and evm prover

rootulp replied:
oh true, would be a good utility.

cmwaters commented on testing/demo/pkg/transfer/relay.go:
```diff
@@ -0,0 +1,261 @@
+
+// FungibleTokenPacketData represents the ICS20 token transfer data structure
+// It should match IICS20TransferMsgs.FungibleTokenPacketData in the Solidity contract
+type FungibleTokenPacketData struct {
```
This isn't already somewhere we can re-use?

rootulp replied:
There is a Go struct for it https://github.com/rootulp/ibc-go/blob/ac0297b3f35e33076732afeedbd3a5dd8f07cc53/modules/apps/transfer/types/packet.pb.go#L26-L40

But there isn't a Go struct for the representation the Solidity expects.

cmwaters commented on testing/demo/pkg/transfer/relay.go:
```diff
@@ -0,0 +1,261 @@
+// relayByTx implements the logic of an IBC relayer.
+// It processes source tx, extracts IBC events, generates proofs,
+// and creates an Ethereum transaction to submit to the ICS26Router contract.
+func relayByTx(sourceTxHash string, targetClientID string) error {
```
ByTx you specifically mean relaying a MsgTransfer right? And this is just from Celestia to ethereum?

rootulp replied:
Yep exactly. This is specific to a MsgTransfer in the Celestia to Ethereum direction.

I can clarify that in comments.

cmwaters commented on testing/demo/pkg/transfer/relay.go:
```diff
@@ -0,0 +1,261 @@
+	}
+	fmt.Printf("SendPacket event: %v\n", event)
+
+	resp, err := getCelestiaProverResponse(event)
```
I would mention this is the state membership proof not the STF.

Also I feel like for safety, we should pass in the latest app hash we updated the tendermint client to to make sure that the membership proof is with reference to that state root. I think it's fine now because no other transactions are submitted but I worry due to the lack of atomicity that this may generate a membership proof up to a different state root

cmwaters replied:
Ah I see that the prover itself handles this by having the client ID passed through in the request

rootulp replied:
Yep exactly. It was difficult to query the last trusted height on the EVM contract from Go code but it's easy to do in Rust b/c there are wrappers. So I opted to just move it inside the celestia-prover.

cmwaters commented on testing/demo/pkg/transfer/update_client.go:
```diff
@@ -0,0 +1,135 @@
+		Sp1Proof sp1proof
+	}{
+		Sp1Proof: sp1proof{
+			VKey:         verifierKey,
```
Does every submission of the proof require the verifier key to be submitted?

rootulp replied:
Seems like it

cmwaters reviewed (approved):
Looks good after the first pass. I'm going to try check it out and run it myself

cmwaters commented:
It looks like this PR doesn't actually remove the solidity-ibc-eureka as a submodule - did you want to if we no longer need it as a submodule?

rootulp commented:
> It looks like this PR doesn't actually remove the solidity-ibc-eureka as a submodule - did you want to if we no longer need it as a submodule?

We still need it as a submodule but we no longer need any changes inside of it. So we can use a commit in upstream rather than a commit from my solidity-ibc-eureka fork
